### PR TITLE
[BUGFIX release] Fix Glimmer memory leak

### PR DESCRIPTION
**Pull request by tomdale**
This commit resolves a memory leak (#11501) in apps running Glimmer. The root cause of the issue was that a flag governing whether components/views remove themselves from their parent was being set incorrectly, so that when the code to cleanup a destroyed view ran, it was not removed from its parent’s `childViews` array.

Specifically, three hooks are invoked when a render node is cleared:
1. `env.hooks.willCleanupNode`
2. `Morph#cleanup`
3. `env.hooks.didCleanupNode`

Prior to this commit, `willCleanupNode` would blindly set the owner view’s `isDestroyingSubtree` flag if there was a view set in the `env` (i.e., basically always).

Sidebar on the `isDestroyingSubtree` flag: this flag is used as a performance optimization. If a view is destroyed, we want to remove that view from its parent’s `childViews` array so that garbage collection can happen. However, it is unnecessary to remove child views from the destroyed view’s `childViews` array; the GC will take care of any clean up when the link between the view hierarchy and the destroyed view is severed.

For example, imagine this hypothetical view hierarchy:

```
      A
     / \
    B   C
       / \
      D   E
```

If the render node for view C is destroyed, we need to remove C from A’s array of child views. However, it is unnecessary to remove D or E from C’s child views, because they will be imminently removed by the GC, and temporarily lingering in the child views array does no harm.

We accomplish this by setting the `isDestroyingSubtree` flag on the root-most view in a hierarchy (view A in the example above), which we call the owner view. When the render node for C is destroyed, it removes C from A’s child views array and sets the flag to true. When D and E’s render nodes are destroyed, they see that the flag has already been flipped and do not remove their associated views from C’s child views array.

The memory leak manifested itself when the a render node that did not have a view associated with it was destroyed, and contained a child render node that _did_ have a view associated. For example:

``` handlebars
{{#if foo}}
  {{my-component}}
{{/if}}
```

In this case, the `willCleanupNode` hook would erroneously set the flag without actually moving the `my-component` view, because the render node for the `{{if}}` does not know about the component.

When the cleanup for `{{my-component}}` finally happens (in Morph#cleanup), the flag has already been set, so the render node incorrectly assumes that its view is part of a subgraph that has already been severed.

As far as we can tell, the `willCleanupNode` hook is not doing any cleanup that is not already done in `Morph#cleanup`. We still do need `didCleanupNode` to clear the flag, but can leave individual render node cleanup to the render node itself.

### CODE REVIEW

**Comment by stefanpenner** [comment]
+1 bugfix that kills removes features

File: packages/ember-views/tests/views/view/child_views_test.js [hunk 1]
```diff
@@ -68,3 +69,41 @@ QUnit.test('should not duplicate childViews when rerendering', function() {
...
+QUnit.test('should remove childViews on destroy', function() {
+  var outerView = EmberView.extend({
+    component: 'my-thing',
+    value: false,
+    container: {
+      lookup() {
+        return {
+          componentFor(tagName, container) {
+            return Component.extend({
+              destroy() {
```
Comment (igorT) [inline packages/ember-views/tests/views/view/child_views_test.js]: Why do we need this?

Comment (rwjblue) [inline packages/ember-views/tests/views/view/child_views_test.js]: We should probably add an assertion here, to confirm that `destroy` is called...

Comment (tomdale) [inline packages/ember-views/tests/views/view/child_views_test.js]: Removed, thanks for catching that.

**Comment by stefanpenner** [comment]
i'll be testing this shortly

**Comment by stefanpenner** [comment]
I'm still seeing the original issue, I will debug further. I suspect it may be a second leak, although it feels quite similar.

also should the following remain?:

* https://github.com/tildeio/htmlbars/blob/master/packages/htmlbars-runtime/lib/hooks.js#L1089
* https://github.com/tildeio/htmlbars/blob/7d5338d4a358bebdb89b1b053944c325a410c88a/packages/htmlbars-runtime/tests/main-test.js#L70
* https://github.com/tildeio/htmlbars/blob/7d5338d4a358bebdb89b1b053944c325a410c88a/packages/htmlbars-util/lib/template-utils.js#L131

**Comment by stefanpenner** [comment]
I have verified an additional dominator, retaining a similar set of info. Although this PR appears to have fixed something. There exists at-least an additional issue. I'll reduce another example

**Comment by stefanpenner** [comment]
@tomdale I have added an additional failing test.

**Comment by tomdale** [comment]
@stefanpenner Awesome, thank you. We'll take a look at this first thing tomorrow.

**Comment by jmurphyau** [comment]
@tomdale your PRs are often very detailed.. It's like a blog post around a particular issue/feature.. It's a good read

**Comment by tomdale** [comment]
@jmurphyau That's why I only get 3-4 PRs done per year.

**Comment by stefanpenner** [comment]
let me confirm with the original leak example.

**Comment by stefanpenner** [comment]
Still have a leak. Going to prepare another test after dinner.

**Comment by stefanpenner** [comment]
memory leak Monday comes to an end.

Until next week, Same Bat time, same Bat channel!

**Comment by stefanpenner** [comment]
friend reminder @rwjblue this is a bugfix release.

**Comment by rwjblue** [comment]
@wycats / @tomdale - Please make sure to prefix each commit with '[BUGFIX <channel>]' in the future.

**Comment by raytiley** [comment]
Sweet!

**Comment by tomdale** [comment]
@rwjblue Apologies, I was going to rebase and squash into a single commit but got lazy. I will try to remember to do this in the future.

**Comment by chriscareycode** [comment]
Will this land in 1.13.x , or only in 2.x?

**Comment by rwjblue** [comment]
1.13

**Comment by jakesjews** [comment]
Any chance of this landing in the stable branch soon? This is the only blocker preventing us from merging our conversion to 1.13.

**Comment by mixonic** [comment]
@jakesjews I have no doubt they are working as quickly as possible. Thanks for your patience.

**Pull request merged**